src/Mvc/Mvc.Formatters.Json/src/JsonInputFormatter.cs

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Text;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Formatters.Json;
+
+namespace Microsoft.AspNetCore.Mvc.Formatters
+{
+    /// <summary>
+    /// A <see cref="TextInputFormatter"/> for JSON content.
+    /// </summary>
+    public sealed class JsonInputFormatter : TextInputFormatter, IInputFormatterExceptionPolicy
+    {
+        private readonly JsonSerializerOptions _serializerOptions;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="JsonInputFormatter"/>.
+        /// </summary>
+        /// <param name="options">The <see cref="JsonFormatterOptions"/>.</param>
+        public JsonInputFormatter(JsonFormatterOptions options)
+        {
+            _serializerOptions = options.SerializerOptions;
+
+            SupportedEncodings.Add(UTF8EncodingWithoutBOM);
+            SupportedEncodings.Add(UTF16EncodingLittleEndian);
+
+            SupportedMediaTypes.Add(MediaTypeHeaderValues.ApplicationJson);
+            SupportedMediaTypes.Add(MediaTypeHeaderValues.TextJson);
+            SupportedMediaTypes.Add(MediaTypeHeaderValues.ApplicationAnyJsonSyntax);
+        }
+
+        /// <inheritdoc />
+        InputFormatterExceptionPolicy IInputFormatterExceptionPolicy.ExceptionPolicy => InputFormatterExceptionPolicy.MalformedInputExceptions;
+
+        /// <inheritdoc />
+        public override async Task<InputFormatterResult> ReadRequestBodyAsync(
+            InputFormatterContext context,
+            Encoding encoding)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (encoding == null)
+            {
+                throw new ArgumentNullException(nameof(encoding));
+            }
+
+            var httpContext = context.HttpContext;
+            var inputStream = GetInputStream(httpContext, encoding);
```

Comment:
Do we need to do any work to make this guy check for invalid characters/encoding? I'm not sure if that's something we need or not now that the serializer operates on bytttes

Reply:
That would invalidate the performance wins no? That would mean everyone that uses the serializer needs to do that as well.

cc @GrabYourPitchforks any guidance here?

Reply:
Security is a thing tho.

Reply:
GIGO?

Reply:
Filed https://github.com/aspnet/AspNetCore/issues/8475 so we don't lose track of this

Reply:
I'd say yes, and you'd reject the json as invalid, What was the behaviour of the previous formatters?

Reply:
The behaviour of the previous formatter was that we passed JSON.NET a streamreader and they did what they do. https://github.com/aspnet/AspNetCore/blob/master/src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs#L145

This is a new serializer so we need to know that the expectations we have are correct. 

/cc @JamesNK in case he has SPICY thoughts

Reply:
We can’t have logic to validate the JSON  that’s the deserialzers job, this is about validating UTF8.

Reply:
It's a good question for @ahsonkhan. I'm thinking the reader should handle this case if that isn't already the case today with an option to turn it off as a fast path if you already did the work yourself.

Reply:
The `System.Text.Json` reader itself should be performing this check on your behalf.

Reply:
Newtonsoft.Json had a check for bad escape sequences. Otherwise it assumed that all the char values that TextReader gave it were valid.

Reply:
@ahsonkhan \ @steveharter is this already being done? Do we need a separate issue tracking implementing \ verifying this?

Reply:
cool beans!

Reply:
> is this already being done?

The `Utf8JsonReader.Read()`, and the `JsonDocument.Parse()` on their own validate/reject invalid UTF-8 data as long as it is **not** contained within a JSON string/property name. This means token delimiters, whitespace, etc. all have to be valid UTF-8. Otherwise, as part of JSON validation, these invalid UTF-8 bytes will be detected and rejected (throw `JsonReaderException`). For invalid/ill-formed UTF-8 within quoted strings, if you were to ever materialize the token into a UTF-16 string by calling `Utf8JsonReader.GetString()` or `JsonElement.GetString()`/`GetRawText()`, that will validate and throw. The `Utf8JsonReader` does **not** validate invalid UTF-8 bytes within quoted strings if you never materialize the strings within the reader loop.

```C#
string jsonString = "[[[]], {\"a\":1, \"b\": 2}, 3, {\"a\":{}, \"b\":{\"c\":[]} }, [{\"a\":1, \"b\": 2}, null]]";
byte[] dataUtf8 = Encoding.UTF8.GetBytes(jsonString);

var reader = new Utf8JsonReader(dataUtf8, isFinalBlock: true, default);
var stringIndices = new List<int>();
while (reader.Read())
{
    if (reader.TokenType == JsonTokenType.PropertyName)
    {
        // Move back to after the first quote
        // Before: "<name>":|
        // After: "|<name>":
        stringIndices.Add((int)reader.BytesConsumed - reader.ValueSpan.Length - 2);
    }
}

for (int i = 0; i < dataUtf8.Length; i++)
{
    dataUtf8 = Encoding.UTF8.GetBytes(jsonString);
    dataUtf8[i] = 0xEF;
    if (stringIndices.Contains(i))
    {
        // Invalid UTF-8 bytes contained within quoted strings are not rejected until the string is materialized
        using (JsonDocument doc = JsonDocument.Parse(dataUtf8))
        {
            // navigating to the token/JsonElement with the invalid UTF-8 byte and calling GetString() on it will throw
        }
    }
    else
    {
        Assert.Throws<JsonReaderException>(() => JsonDocument.Parse(dataUtf8));
    }
}
```

The JSON Deserializer APIs (like `JsonSerializer.Parse`) will reject invalid UTF-8 tokens if it occurs anywhere within the JSON payload (including within JSON strings/properties).
```C#
dataUtf8 = Encoding.UTF8.GetBytes("\"hello\"");
Assert.Equal("hello", Serialization.JsonSerializer.Parse<string>(dataUtf8));
dataUtf8[3] = 0xEF;
Assert.Throws<InvalidOperationException>(() => Serialization.JsonSerializer.Parse<string>(dataUtf8));

dataUtf8 = Encoding.UTF8.GetBytes("[ ]");
Assert.Equal(Array.Empty<int>(), Serialization.JsonSerializer.Parse<int[]>(dataUtf8));
dataUtf8[1] = 0xEF;
Assert.Throws<JsonReaderException>(() => Serialization.JsonSerializer.Parse<int[]>(dataUtf8));
```

Reply:
> I'm thinking the reader should handle this case if that isn't already the case today with an option to turn it off as a fast path if you already did the work yourself.

@davidfowl, based on the above clarification, do you still think the **reader** needs to validate UTF-8 (particularly within quoted strings) even if no one materializes them? Currently, that validation is lazy. The **serializer** does the validation already primarily because it needs to get the string eventually.